noxfile.py

```diff
@@ -0,0 +1,388 @@
+# Copyright iris-grib contributors
+#
+# This file is part of iris-grib and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+"""
+Perform test automation with nox.
+
+For further details, see https://nox.thea.codes/en/stable/#
+
+"""
+
+from contextlib import contextmanager
+import hashlib
+import os
+from pathlib import Path
+
+import nox
+import yaml
+
+#: Default to reusing any pre-existing nox environments.
+nox.options.reuse_existing_virtualenvs = True
+
+#: Name of the package to test. 
+PACKAGE = str("iris_grib")
+
+#: Cirrus-CI environment variable hook.
+PY_VER = os.environ.get("PY_VER", ["3.6", "3.7", "3.8"])
+IRIS_SOURCE = os.environ.get("IRIS_SOURCE", ['source', 'conda-forge'])
+
+#: Default cartopy cache directory.
+CARTOPY_CACHE_DIR = os.environ.get("HOME") / Path(".local/share/cartopy")
+
+
+def venv_cached(session, requirements_file=None):
+    """
+    Determine whether the nox session environment has been cached.
+
+    Parameters
+    ----------
+    session: object
+        A `nox.sessions.Session` object.
+
+    Returns
+    -------
+    bool
+        Whether the session has been cached.
+
+    """
+    result = False
+    if not requirements_file is None:
+        yml = Path(requirements_file)
+    else:
+        yml = Path(f"requirements/ci/py{session.python.replace('.', '')}.yml")
+    tmp_dir = Path(session.create_tmp())
+    cache = tmp_dir / yml.name
+    if cache.is_file():
+        with open(yml, "rb") as fi:
+            expected = hashlib.sha256(fi.read()).hexdigest()
+        with open(cache, "r") as fi:
+            actual = fi.read()
+        result = actual == expected
+    return result
+
+
+def concat_requirements(primary, *others):
+    """Join together the dependencies of one or more requirements.yaml.
+
+    Parameters
+    ----------
+    primary: str
+        filename of the primary requirements.yaml
+
+    others: list[str]
+        list of additional requirements.yamls
+    
+    Returns
+    -------
+    yaml
+        Dictionary of yaml data: primary with the addition
+        of others[]['dependencies']
+
+    """
+    with open(primary, 'r') as f:
+        requirements = yaml.load(f, yaml.FullLoader)
+
+    for o in others:
+        with open(o, 'r') as f:
+            oreq = yaml.load(f, yaml.FullLoader)
+            requirements['dependencies'].extend(oreq['dependencies'])
```

Comment:
@jamesp I guess this is okay coping with duplicate packages from different requirements?

Reply:
Yes this is ok with duplicate entries.  I guess we could put a `set` around it if we want to be explicit.

If the dependencies are incompatible (e.g. versions not resolvable), conda will fail to build the environment.  Example below. I expect this is what we want - we do not want iris and iris-grib dependencies to diverge to the point that they are irresolvable.

```
jpenn@vld159:~  $ cat requirements.yml 
name: _test

channels:
  - conda-forge

dependencies:
  - python=3.8
  - requests
  - nose
  - nose>=1.3
  - nose>1.2
jpenn@vld159:~  $ conda env create --file=requirements.yml
Collecting package metadata (repodata.json): done
Solving environment: done
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate _test
#
# To deactivate an active environment, use
#
#     $ conda deactivate

jpenn@vld159:~  $ conda env remove -n _test

Remove all packages in environment /home/h06/jpenn/local/miniconda3/envs/_test:

jpenn@vld159:~  $ echo "  - nose<1.1" >> requirements.yml 
jpenn@vld159:~  $ conda env create --file=requirements.yml
Collecting package metadata (repodata.json): done
Solving environment: failed

ResolvePackageNotFound: 
  - nose[version='<1.1,>1.2,>=1.3']
```

Reply:
This is fine as it stands :+1: